### PR TITLE
[DevTools] Browser dev tools should open default browser

### LINKED ISSUE #492: firefox devtools
What would it take to support Firefox Devtools as well? I suspect that facebook/react-devtools#56 may be a blocker.

### PR DESCRIPTION
**Pull request by iamdustan**
This changes the browser dev tools tunneling to open the default browser using *nix’s `open` command instead of applescript.

Benefits:
* leave browser tooling decisions to the user.
* linux support

Limitations:
* Doesn’t detect whether the tab is already open and focus if so (like the Chrome applescript implementation did).

I think that limitation needs to be tackled before merging. I’m not really sure where to start to tackle that. I suspect this is low priority to your team, but is this something you’d be interested in?

Closes #492

### CODE REVIEW

**Comment by frantic** [comment]
One possible solution is to make the websocket-proxy smarter. Instead of just broadcasting incoming messages it should be able to detect browser connection and simulator connection and pair them. If there is incoming simulator connection without existing browser counterpart, proxy should open a new browser tab.

**Comment by iamdustan** [comment]
Great idea, @frantic. I’ll explore that over the next day or two.

**Comment by iamdustan** [comment]
@frantic this was a bit of a hack, but there is already a comment saying “this is hacky, refresh the browser window to reattach”.

Originally the chrome already open or not is happened in AppleScript. This just sets a local variable on initial browser connection and then bails out in the future.

I've tested it in chrome and fx dev edition on my bus ride and it was working well for me.

File: packager/packager.js [hunk 1]
```diff
@@ -127,23 +127,30 @@ function openStackFrameInEditor(req, res, next) {
...
-      res.end('OK');
+    } else if (req.url === '/launch-browser-devtools') {
+      if (connected) {
+        console.log('Already connect to Dev Tools in default browser...');
+        res.end('OK');
+      }
+      else {
+        var debuggerURL = 'http://localhost:' + options.port + '/debugger-ui';
+        console.log('Launching Dev Tools in default browser...');
+        connected = true;
```
Comment (frantic) [inline packager/packager.js]: this means it will open the tab only once per server session.

Comment (iamdustan) [inline packager/packager.js]: Yeah, but that’s same as it is currently (as far as I could tell).

E.g. Once you open and attempt to close or refresh the browser tab it says, “if you do this you’ll break the bridge”

Comment (frantic) [inline packager/packager.js]: But when you close the tab and disable debugging on simulator (via Cmd+Ctrl+Z) or restart the app in Xcode, it's important that the tool can connect to the debugger again.

Comment (iamdustan) [inline packager/packager.js]: So the two use cases I’m currently missing are:

1. Cmd+Ctrl+Z switches back to running js on the device. This should reset the connected state so future Cmd+D can reconnect to a browser tab.
2. Restarting the simulator from XCode without closing the window should be able to reconnect.

Am I missing anything else?

**Comment by nicklockwood** [comment]
This seems problematic because we use different executor implementations for Chrome/Safari (and don't support Firefox) - unless I'm missing something obvious?

**Comment by iamdustan** [comment]
@nicklockwood is this something that has changed over the past month since I opened this? At the time simply having it open my default browser worked fine.

Still need to work in some time to come revisit this and finish it.

Ignore the [GH Review: review-needed] label and set it back to needs-revision. I’ve only rebased master, haven’t done any additional work yet.

**Comment by brentvatne** [comment]
@iamdustan - any interest in continuing work on this or should we close?

**Comment by iamdustan** [comment]
@brentvatne I am interested, but am currently on 3 projects at work so my time is a bit scarce. :/

Feel free to close for now to get it off the backlog and someday I’ll make it back around to this.

**Comment by sahrens** [comment]
closing for now - feel free to re-open when it's ready for review - we'll need to make sure our full tool stack actually works in chrome, safari, and firebox before we enable something like this, though, otherwise it's just going to open in a broken environment.

**Pull request closed without merging**